### ISSUE TITLE
Discuss - 0036-launcher-api-v0.5

### ISSUE THREAD

**Issue by rossmuir** [maidsafe/rfcs#141]
RFC > https://github.com/maidsafe/rfcs/tree/master/text/0036-launcher-api-v0.5

**Comment by theWebalyst** [comment]
What's the status of file version support in NFS? 

I'd like to know what is planned, or if not planned then whether this is to be addressed soon or not really being considered, or...  etc because: remoteStorage! :-)

There is a flag in the NFS API to say if a directory should be versioned, but I'm not aware of any API that builds on this. (I haven't checked 0.5 but there's no mention here so I assume that has not changed. )

**Comment by cretz** [comment]
Notes:

1. "Rename custom headers to start with X- as mentioned in the RFC" <-- You are reading the wrong RFC. They have deprecated this: https://tools.ietf.org/html/rfc6648.
1. For file things (and the rest of the API), I personally would love to either see the parameters switched or move isPathShared to some kind of header. For instance if they are switched, i.e. /nfs/file/private/my/file/path.txt and /nfs/file/shared/my/file/path.txt, you don't have to encode the file path which is nice and clean (has a nice WebDAV feel).
1. "Remove the custom headers because the metadata request (HEAD) can be used to fetch the file metadata and restrict the GET file requests only for reading the content of the file." <-- actually you should give the same headers in my opinion (maybe with the exception of the metadata headers, maybe make that an option). There should be no harm in providing this and it's reasonable to want the created/modified information on a GET along with `Accept-Ranges: bytes`, etc. Again, I can understand avoiding the metadata though (though an option would be nice on the GET, not sure how you'd do it cleanly)
1. "and the data is appended to the end while writing." <-- There is no good sensible default here whether to start at byte 0 or append by default...so this is fine, just make it really clear in the docs because it can seem counterintuitive
1. For file read mime type, I think it needs to be made clear that you are just guessing based on filename...no real problem here because it's not a significant performance penalty, but I know I probably won't be using the value
1. For "Streaming Issue in the Web Platform" I think you're a bit off. I don't think you need multipart HTTP file upload. Just process a normal PUT and read the bytes as the file. This works just fine and streams just fine. Remember, streaming is really just about how bytes are handled on either end, not really about some special protocol support (as opposed to range headers which I'd argue is more about "chunking"). So just stream the HTTP body of my PUT/POST as the file data and don't worry about multipart
1. In alternatives, I think yyou're misunderstanding HTTP streaming here. Web sockets don't help streaming any more than just processing an HTTP body as a stream. Web sockets help with bidirectional communication. You will not stream a file any faster/better over web socket than you would using a normal HTTP request w/ the bytes as the body. Remember, when developed properly, the HTTP body I would send you when uploading a file I won't have all in memory (nor would you), I would be streaming it while the request is ongoing.
1. Concerning headers vs body on metadata response...meh, your call. Though you might find sending the metadata in the body to be easier because you don't have to base 64 it, but since you're using `HEAD` we should continue to avoid a body altogether. If you changed the HTTP method, then yes, putting metadata in the body instead of the header is very reasonable. I think the current way is fine unless there is fear of really large metadata.

**Comment by krishnaIndia** [comment]
@cretz, thanks for all the notes :+1: 

> "Rename custom headers to start with X- as mentioned in the RFC" <-- You are reading the wrong RFC. They have deprecated this: https://tools.ietf.org/html/rfc6648.

Thanks for pointing it out. I will get this updated in the RFC.

> For file things (and the rest of the API), I personally would love to either see the parameters switched or move isPathShared to some kind of header. For instance if they are switched, i.e. /nfs/file/private/my/file/path.txt and /nfs/file/shared/my/file/path.txt, you don't have to encode the file path which is nice and clean (has a nice WebDAV feel).

Switching it is fine and that makes the parameters to be mandatory. But that seems to be worthy to switch.

> "Remove the custom headers because the metadata request (HEAD) can be used to fetch the file metadata and restrict the GET file requests only for reading the content of the file." <-- actually you should give the same headers in my opinion (maybe with the exception of the metadata headers, maybe make that an option). There should be no harm in providing this and it's reasonable to want the created/modified information on a GET along with Accept-Ranges: bytes, etc. Again, I can understand avoiding the metadata though (though an option would be nice on the GET, not sure how you'd do it cleanly)

I do agree

> "and the data is appended to the end while writing." <-- There is no good sensible default here whether to start at byte 0 or append by default...so this is fine, just make it really clear in the docs because it can seem counter intuitive

Will try to update in the document to make it clear

>For "Streaming Issue in the Web Platform" I think you're a bit off. I don't think you need multipart HTTP file upload. Just process a normal PUT and read the bytes as the file. This works just fine and streams just fine. Remember, streaming is really just about how bytes are handled on either end, not really about some special protocol support (as opposed to range headers which I'd argue is more about "chunking"). So just stream the HTTP body of my PUT/POST as the file data and don't worry about multipart

> In alternatives, I think you're misunderstanding HTTP streaming here. Web sockets don't help streaming any more than just processing an HTTP body as a stream. Web sockets help with bidirectional communication. You will not stream a file any faster/better over web socket than you would using a normal HTTP request w/ the bytes as the body. Remember, when developed properly, the HTTP body I would send you when uploading a file I won't have all in memory (nor would you), I would be streaming it while the request is ongoing.

While receiving data from the server, the `XHR onprogress` event can be used to read the data as and when it is received. When we try to upload a file, the file content can be sent as you said using the PUT/POST.  But when we have to upload a huge file, then we must read the file in smaller range using the FileReader's `slice` method and upload the file in smaller chunks to the server.

In our case uploading a file in smaller chunks was resulting in large number of chunks generated by the SelfEncryptor. The append operation seems to create additional chunks, so the ideal usage of the self encryptor is to obtain the reference, stream the data and finally call the close function of the self encryptor.

Even if we want to support chunked uploads then we must probably take the route like what few Cloud Storage providers do.
First hit a URL and get a ID referring to the upload and then upload data in chunks by passing the ID. Once the content size specified is reached the data gets stored (something like that).

This approach is similar to the alternative suggested, but the only difference is using many short lived connections for chunked uploads vs single long lived connection using websocket.

The two options that I could find was to upload in chunks and the other was to use a Websocket connection. I do agree websocket is just an other long lived HTTP connection and does not have significant performance improvement. But the only reason for opting the Websocket approach was to avoid creating too many short lived connections and even if there are multiple file uploads a SharedWorker can be created and files can be uploaded using a single connection.

I might be missing something as you say, but I am eager to know If there is an better work approach that can be added to the RFC instead of this.

> Concerning headers vs body on metadata response...meh, your call. Though you might find sending the metadata in the body to be easier because you don't have to base 64 it, but since you're using HEAD we should continue to avoid a body altogether. If you changed the HTTP method, then yes, putting metadata in the body instead of the header is very reasonable. I think the current way is fine unless there is fear of really large metadata.

As of now we don't have a cap on metadata. But for sure it will be restricted in the near future.

**Comment by krishnaIndia** [comment]
@theWebalyst am looking into the version support. Thanks for brining it up. Will keep you posted here on the progress.

**Comment by cretz** [comment]
Thanks for the response.

> While receiving data from the server, the XHR onprogress event can be used to read the data as and when it is received. When we try to upload a file, the file content can be sent as you said using the PUT/POST. But when we have to upload a huge file, then we must read the file in smaller range using the FileReader's slice method and upload the file in smaller chunks to the server.

> In our case uploading a file in smaller chunks was resulting in large number of chunks generated by the SelfEncryptor. The append operation seems to create additional chunks, so the ideal usage of the self encryptor is to obtain the reference, stream the data and finally call the close function of the self encryptor.

> Even if we want to support chunked uploads then we must probably take the route like what few Cloud Storage providers do.
First hit a URL and get a ID referring to the upload and then upload data in chunks by passing the ID. Once the content size specified is reached the data gets stored (something like that).

> This approach is similar to the alternative suggested, but the only difference is using many short lived connections for chunked uploads vs single long lived connection using websocket.

> The two options that I could find was to upload in chunks and the other was to use a Websocket connection. I do agree websocket is just an other long lived HTTP connection and does not have significant performance improvement. But the only reason for opting the Websocket approach was to avoid creating too many short lived connections and even if there are multiple file uploads a SharedWorker can be created and files can be uploaded using a single connection.

> I might be missing something as you say, but I am eager to know If there is a better work approach that can be added to the RFC instead of this.

There is some real confusion here. I am not sure I get what you are saying. XHR/Ajax has nothing to do with any of this. You have a web server API that accepts a big bunch of bytes in a PUT. You accept it and continually stream the data from me over HTTP like any other content. You feed it over the safe FFI however it is best for y'all, but you never hold all of the data in memory. There should be no need for chunked uploads or any of that. Node has great support for streaming in a series of bytes. How y'all choose to self encrypt that and store it (be it a create then a series of appends or whatever) should have nothing to do with the API.

Basically ask yourself: "How would I write it in Rust to store a 5G file from disk to the safe network?" Then assume the IO stream that is from disk is now instead from the HTTP API (and whatever FFI construct y'all use). This isn't really a two step process from the side of the API. There is no web sockets or chunking or any of that. Basically every setup that accepts large content over HTTP streams it out to their backend. I apologize if I am not clear here.

**Comment by krishnaIndia** [comment]
> There is some real confusion here.

Yes, probably I didn't make it clear.

 I was trying to explain how the javascript client running on the browser can upload a huge file to the server.

This [article](http://www.html5rocks.com/en/tutorials/file/xhr2/#toc-example-slicingfiles) explains how large uploads can be done from a browser client. Few other articles also suggest the same two approaches for the web based clients. 

If that is the case, then how can we provide an option to handle large data uploads better. Thus, I also added an alternate plan for exposing an **additional API**  which can help the web clients running on the browser to use that for uploading large data.

Other platforms Rust, Go, C#, Java, nodejs, etc definitely can use the streams to write the data without any problem. But for the clients running on the browser, I could find only these two as the options to handle large data. My point above was just confined to this and exposing an additional api to facilitate the web clients. If there is a better approach for web clients then we don't have to cater for the additional APIs.

**Comment by cretz** [comment]
IMO you should be very careful altering the API to suit a particular use case. I personally will never use it from a browser. I also am 1 bazillion % against having the API visible from the normal user browser (I have opened [an issue](https://maidsafe.atlassian.net/browse/CS-62) on this and definitely consider the launcher to be incredibly insecure until it is removed).

In my opinion, if someone wants a limited environment like a browser to be able to upload files, they need the assistance of a separate app which can proxy browser-based requests and make it easier. IMO you should not go down the slippery slope of supporting very specific use cases. The API should be generic.

Having said that, using the file APIs is the only way that JS in a web browser can access an uploaded file. And AJAX does not provide a way to stream data to a URL (and I am not aware of another native way). But once the extreme insecurity of opening the launcher up to the world is resolved, who would be accessing the launcher directly through a browser? Even Electron apps have node right there.

**Comment by krishnaIndia** [comment]
@theWebalyst, we are holding off the version support for now. Enabling versioning can increase the data volume stored and also spike up network traffic too. It would be better to release this feature at a later point of time once we have implemented few concepts like `Deleting Owned Immutable data` and also improved client caching to reduce network traffic.

Will update the RFC with few more points and also to reflect the changes that I have agreed with @cretz

**Comment by rossmuir** [comment]
RFC Updated as per PR > https://github.com/maidsafe/rfcs/pull/145

**Comment by cretz** [comment]
Notes on latest changes:

1. "The FFI errors are sent to caller as 500 Status code in few cases." <-- that's too bad. Should only be a 500 if it's a server-side problem.
1. "Regex for validating the service and public name ^[a-zA-Z0-9][a-zA-Z0-9-]{1,61}[a-zA-Z0-9]+$" - This appears to be a hostname validation, right? Seems reasonable...but I am confused by the 61 limit considering there is an unlimited repetition check after it...are you just trying to make sure I only have up to 61 dashes (i.e. `-` which you need to escape there anyways)
1. I completely disagree with "CORS validate". CORS has no place in a normal HTTP API unless you want it accessed by a browser, which I completely disagree with.
1. Why have CSP headers on an API? It doesn't really do anything unless the API's response would be rendered as HTML by a browser...are you confusing this with the web proxy?
1. You should make things like NFS create directory use the path in the URL instead of providing it via JSON (sure, the metadata can still be provided there or something)...it is more restful
1. Why did you just make the `isPathShared` a boolean? `/nfs/directory/true/mydir` or `/nfs/directory/false/mydir` makes little sense compared to `/nfs/directory/shared/mydir` or `/nfs/directory/private/mydir`
1. Get dir says it returns a base64'd set of metadata, but create dir doesn't say anything about base 64 of the metadata...I assume that it is a base 64'd set of bytes on create dir?
1. Updating file metadata should not be at `/nfs/file/` where a `PUT` would make someone think they are creating a file...this should probably be `/nfs/filemetadata/` or something, I dunno
1. So "Update File Metadata" also allows a name change?
1. "Thus the NFS file content upload API must be able to support multipart upload." <-- no, there is another option: no browser access to the API. Sadly this is not being considered.

**Comment by krishnaIndia** [comment]
> Should only be a 500 if it's a server-side problem.

Even I meant the same. I ll rephrase it in the document if it is not clear.

> are you just trying to make sure I only have up to 61 dashes

I will test this and update the regex

> Why have CSP headers on an API? It doesn't really do anything unless the API's response would be rendered as HTML by a browser...are you confusing this with the web proxy?

It is my mistake again there.

> You should make things like NFS create directory use the path in the URL instead of providing it via JSON (sure, the metadata can still be provided there or something)...it is more restful

I see an advantage in this approach that you have suggested. Having the standard end point consistently would be great. If the devs have to add the metadata, it can just be sent as a JSON object. JSON object would be better because it would be easy to scale (considering the version option to be provided in future). And in this approach we also make all the mandatory fields as a part of the URL itself and the body of the request becomes completely optional.

> Why did you just make the isPathShared a boolean? /nfs/directory/true/mydir or /nfs/directory/false/mydir makes little sense compared to /nfs/directory/shared/mydir or /nfs/directory/private/mydir

Making it an enum seems to be better. Thanks for this suggestion

> Get dir says it returns a base64'd set of metadata, but create dir doesn't say anything about base 64 of the metadata...I assume that it is a base 64'd set of bytes on create dir?

Yes, it should be a base 64'd set of bytes.

> Updating file metadata should not be at /nfs/file/ where a PUT would make someone think they are creating a file...this should probably be /nfs/filemetadata/ or something, I dunno

How about using `PATCH` verb for updating the name/metadata, for directory and also for file. So we don't have to change the endpoints and using a similar end point structure through out. 

> So "Update File Metadata" also allows a name change

Yes, we do allow the update of Name and metadata. Probably the title I gave is misleading.

**Comment by cretz** [comment]
> How about using PATCH verb for updating the name/metadata, for directory and also for file. So we don't have to change the endpoints and using a similar end point structure through out.

I think updating metadata and updating a file should be so distinctly different as to warrant different paths

**Comment by rossmuir** [comment]
RFC updated as per PR > https://github.com/maidsafe/rfcs/pull/147

**Comment by cretz** [comment]
LGTM

I suggest removing `srcRootPath` and `destRootPath` and instead just having `/app` or `/drive` prefix the `srcPath` and `destPath`. I think it makes things clearer, but meh, your call. I also do NOT think it should ever be optional (it isn't when it's in the URL).

And of course I'll just restate for the record that I am against the web proxy and I am against use of the API from browsers.

**Comment by afck** [comment]
Implemented.

**Comment by loziniak** [comment]
Hi. I get errors in my browser about your CORS headers. I think you should enclose self and none keywords in single quotes like this:
`default-src 'self' *.safenet; object-src 'none'; base-uri 'self';    form-action http://api.safenet; frame-ancestors 'self';X-Frame-Options : SAMEORIGIN`
More, should "X-Frame-Options : SAMEORIGIN" really be a part of CORS header value? I doubt.
cheers :-)

**Comment by krishnaIndia** [comment]
Thanks @loziniak. Here is the corrected CSP header,
`default-src 'self' *.safenet; object-src 'none'; base-uri 'self'; form-action http://api.safenet; frame-ancestors *.safenet; child-src *.safenet`

This will be reflected in the master branch with the next version update
